lib/datasource/terraform-module/index.ts

```diff
@@ -1,112 +1,66 @@
-    cacheNamespace,
-    registryUrl,
-    serviceDiscovery,
-    cacheMinutes
-  );
+  readonly defaultVersioning = hashicorpVersioning.id;
 
-  return serviceDiscovery;
-}
-/**
- * terraform.getReleases
- *
- * This function will fetch a package from the specified Terraform registry and return all semver versions.
- *  - `sourceUrl` is supported of "source" field is set
- *  - `homepage` is set to the Terraform registry's page if it's on the official main registry
- */
-export async function getReleases({
-  lookupName,
-  registryUrl,
-}: GetReleasesConfig): Promise<ReleaseResult | null> {
-  const { registry, repository } = getRegistryRepository(
+  /**
+   * This function will fetch a package from the specified Terraform registry and return all semver versions.
+   *  - `sourceUrl` is supported of "source" field is set
+   *  - `homepage` is set to the Terraform registry's page if it's on the official main registry
+   */
+  @cache({
+    namespace: `datasource-${TerraformModuleDatasource.id}`,
+    key: (getReleasesConfig: GetReleasesConfig) =>
+      TerraformModuleDatasource.getCacheKey(getReleasesConfig),
+  })
+  async getReleases({
     lookupName,
-    registryUrl
-  );
-  logger.debug(
-    { registry, terraformRepository: repository },
-    'terraform.getDependencies()'
-  );
-  const cacheNamespace = 'terraform-module';
-  const cacheURL = `${registry}/${repository}`;
-  const cachedResult = await packageCache.get<ReleaseResult>(
-    cacheNamespace,
-    cacheURL
-  );
-  // istanbul ignore if
-  if (cachedResult) {
-    return cachedResult;
-  }
-  try {
-    const serviceDiscovery = await getTerraformServiceDiscoveryResult(
-      registryUrl
+    registryUrl,
+  }: GetReleasesConfig): Promise<ReleaseResult | null> {
+    const { registry, repository } =
+      TerraformModuleDatasource.getRegistryRepository(lookupName, registryUrl);
+    logger.debug(
+      { registry, terraformRepository: repository },
+      'terraform-module.getReleases()'
     );
-    const pkgUrl = `${registry}${serviceDiscovery['modules.v1']}${repository}`;
-    const res = (await http.getJson<TerraformRelease>(pkgUrl)).body;
-    const returnedName = res.namespace + '/' + res.name + '/' + res.provider;
-    if (returnedName !== repository) {
-      logger.warn({ pkgUrl }, 'Terraform registry result mismatch');
-      return null;
+
+    let res: TerraformRelease;
+    let pkgUrl: string;
+
+    try {
+      const serviceDiscovery = await this.getTerraformServiceDiscoveryResult(
+        registryUrl
+      );
+      pkgUrl = `${registry}${serviceDiscovery['modules.v1']}${repository}`;
+      res = (await this.http.getJson<TerraformRelease>(pkgUrl)).body;
+      const returnedName = res.namespace + '/' + res.name + '/' + res.provider;
+      if (returnedName !== repository) {
+        logger.warn({ pkgUrl }, 'Terraform registry result mismatch');
+        return null;
+      }
+    } catch (err) {
+      this.handleGenericErrors(err);
```

Comment:
```suggestion
      this.handleGenericErrors(err);
      return null;
```
Break early here? Otherwise next steps will do strange things. I think `handleGenericErrors` did not always throw?

Reply:
The signature for this is `handleGenericErrors(err: HttpError): never` as the last line is `throw err`. `return null` would never get called.

What we really need is https://github.com/microsoft/TypeScript/issues/13219